IOS-5515 Add TransactionValidator

ba1ashov opened the pull request:
To be checked: 
- [x]  Default implementation
- [x] DustRestrictable
- [x] MinimumBalanceRestrictable
- [x] WithdrawalValidator
- [x] ReserveAmountRestrictable

ba1ashov committed:
IOS-5515 Add TransactionValidator

ba1ashov committed:
IOS-5515 remove no `async`

IOS-5515 Add validate method

ba1ashov committed:
IOS-5515 Fix selfreview

IOS-5515 Update TransactionValidator.swift

Balashov152 commented on BlockchainSdk/Common/Interfaces/TransactionCreator.swift:
```diff
@@ -8,7 +8,6 @@
 
 import Foundation
 
-@available(iOS 13.0, *)
```
можно завести задачку, это удалить везде, так у нас уже target - 14.5

Balashov152 commented on BlockchainSdk/Common/Interfaces/TransactionValidator/TransactionValidator.swift:
```diff
@@ -0,0 +1,136 @@
+import Combine
+import TangemSdk
+
+public protocol TransactionValidator: WalletProvider {
+    func validate(amount: Amount, fee: Fee, destination: DestinationType?) async throws
+    func validate(amount: Amount, fee: Fee) throws
+}
```
Идея в том, что бы был один `TransactionValidator` и через дефолтные экстеншены была реализация. 

В идеале бы один метод сделать, который `async`, что можно было только через него все валидировать. Но пока не знаю не овер ли это. Так как это нужно только для запроса `isAccountCreated`.

tureck1y replied:
все же нельзя давать создавать транзу без валидации

Balashov152 replied:
Добавил еще метод в WalletManger. Теперь если он поддерживает валидацию, то создание будет через валидацию проходить

Balashov152 commented on BlockchainSdk/Common/BaseManager.swift:
```diff
@@ -17,10 +17,6 @@ class BaseManager: WalletProvider {
     
     var cardTokens: [Token] = []
-    
-    var defaultSourceAddress: String { wallet.address }
-    var defaultChangeAddress: String { wallet.address }
```
Перенес в `extension WalletProvider`

Balashov152 commented on BlockchainSdk/Common/Interfaces/TransactionValidator/ValidationError.swift:
```diff
@@ -0,0 +1,52 @@
+
+import Foundation
+
+public enum ValidationError: Hashable, LocalizedError {
```
Большинство ошибок с `TransactionError`

ba1ashov committed:
IOS-5515 Add `no async` method

ba1ashov committed:
IOS-5515 Add default

ba1ashov committed:
IOS-5515 Update AnyPublisher+Async.swift

ba1ashov committed:
IOS-5515 Update BlockchainSdkTests.swift

ba1ashov committed:
IOS-5515 Update TransactionValidator.swift

ba1ashov committed:
IOS-5515 Merge branch 'develop' into IOS-5515-Refactoring-validation

tureck1y commented on BlockchainSdk/Common/Interfaces/TransactionCreator.swift:
```diff
@@ -32,8 +27,8 @@ public extension TransactionCreator {
         contractAddress: String? = nil
-    ) throws -> Transaction {
-        try self.createTransaction(
+    ) -> Transaction {
```
давай обсудим, не очень понятно стало

tureck1y commented on BlockchainSdk/Common/Interfaces/TransactionValidator/ReserveAmountRestrictable.swift:
```diff
@@ -0,0 +1,19 @@
+
+enum ReserveAmountRestrictableAddressType {
+    case address(String)
+    case new
```
а как это использовать?

Balashov152 replied:
Оптимизация, что если адрес не указан, что бы запрос не отправлять

Balashov152 replied:
Добавил комменты ✅

tureck1y commented on BlockchainSdk/Common/Interfaces/TransactionValidator/TransactionValidator.swift:
```diff
@@ -0,0 +1,165 @@
+}
+
+public enum DestinationType: Hashable {
+    case generate
+    case address(String)
```
комменты бы не помешали, непонятно что это значит

Balashov152 replied:
Добавил комменты ✅

ba1ashov committed:
IOS-5515 Fix review

ba1ashov committed:
IOS-5515 `TransactionCreator` as `TransactionValidator` inheritor

IOS-5515 Update TransactionCreator.swift

IOS-5515 Update TransactionCreator.swift

ba1ashov committed:
Merge branch 'develop' into IOS-5515-Refactoring-validation

ba1ashov committed:
IOS-5515 Merge branch 'develop' into IOS-5515-Refactoring-validation

ba1ashov commented:
Конфликты поправил
